accepted/2021/experimental-package-shading.md

```diff
@@ -0,0 +1,59 @@
+# Experiment: NuGet Package Shading
+
+(intro and scenarios to be added)
+
+## Usage
+
+### Package authors
+
+A *package author* may enable shading on any package reference as long as its assets are configured such that it is effectively private. Any other package references that have a shaded package reference as a transitive dependency must also be shaded. When the project is restored, shaded copies of the packages are created and swapped in. When the project is packed, the shaded assets are embedded in the resulting package.
+
+### Package consumers
+
+Shading is superficially invisible to consumers of a package that has shaded dependencies. They may notice the shaded assets in the build logs or output directory, and their app size may increase, but they do not need to be aware that shading exists. From the consumer's perspective, a shaded dependency is no different than any other private asset.
+
+## Behavior
+
+### Private assets
+
+This shading mechanism is built upon the existing NuGet concept of *private assets*. A shaded asset is inherently private and must be treated the same way as other private assets. Handling shaded assets using the private assets mechanisms reuses existing behavior and concepts, and should result in a consistent experience.
+
+Shading may be enabled on any package reference that is effectively a *private package reference* i.e. all assets consumed from the package are consumed as private assets, and the reference is not exposed transitively. A future proposal will explore making private package references a first class concept.
+
+### Shaded packages
+
+To ensure consistency in package assets, shading applies to an entire package rather than individual assets or asset types. The core of this feature is a mechanism that can take a package and synthesize a *shaded package* by copying the original package and renaming the copy it and its assets to have a new identity: the *shaded name*.
+
+The shaded name is a mangled name specific to the context in which the package is shaded and is designed such that the shaded package's assets do not collide at runtime with assets from other shaded copies and from the original unshaded package.
+
+The name mangling is an implementation detail, and although it is deterministic to allow for deterministic builds, it may change in future versions of the .NET SDK. Developers should not depend on the mangling format or specific mangled IDs outside of references created by the .NET SDK itself.
+
+When a shaded package has a dependency on a package that is also shaded, its assets will be processed so that any references to the dependency's unshaded assets are updated to reference the shaded assets.
+
```

Comment:
Looks like we need a couple of examples to explain this better.

Reply:
Added examples of transitive shading